### DISCUSSION TITLE
Unified module style?

### DISCUSSION THREAD

**Discussion by Kixunil** [rust-bitcoin/rust-bitcoin#1446] [Refactoring]
Basically `foo/mod.rs` vs `foo.rs`. I vote for the latter because it makes adding submodules easier (for both humans and git) but it's not something I'm too annoyed about.

**Comment by tcharding** [comment]
Ha! I've been waiting for this to come up since I started hacking on `rust-bitcoin` crates. On a previous project I pushed hard to move from `mod.rs` to the newer `foo.rs` and a few months later realised that it made zero difference to my day - there are pros and cons of each.

I think it improves the experience for contributors if we use the modern style though, its less jarring.

FTR the change is trivial and can be done with a shell script. Did a quick and dirty test and git can handle rebasing on top of module moves so I guess this is non-invasive apart from everyone having to remember things have moved.

**Reply by Kixunil** [reply]
What are the cons of newer style? I don't see any besides people not being used to it.

> everyone having to remember things have moved

Yeah, we should have a clippy lint as well to avoid inconsistency.

**Comment by tcharding** [comment]
The pros/cons I was referring to are:
- con: different editor keystrokes to open a file in the same directory vs up a directory
- pro: less name clashes in buffers (i.e., now one often has multiple `mod.rs` files open)

These are super minor - I'm totally in favour of using the new style.

**Reply by dpc** [reply]
Another tiny, tiny con for new style is that moving a module is moving one file + one dir, and not just one dir.

I also prefer the new style, mostly because of file name being better for fuzzy search, and when displayed in various places. Tens of `mod.rs` files being mixed together is sometimes annoying.

**Reply by Kixunil** [reply]
Good points. Though I don't `cd` into every directory like many people do. For Rust projects I only `cd` into crate root, so `mod.rs` is actually *more* keystrokes form me. (But I tab the shit out of it anyway.)

**Comment by apoelstra** [comment]
The suggestion here is that we'd have `foo.rs` and `foo/` with submodules in it? That would greatly reduce ergonomics because now `f<tab>` will only autocomplete `foo`, requiring an additional `.<tab>` to get `foo.rs` (same as now, except you do `m<tab>` rather than `.<tab>`) and an additional `/b<tab>` to get `foo/bar.rs` (which is an extra `/` character, both surprising and far away).

I really don't feel strongly but having directories that are prefixes of filenames is a pretty annoying papercut and I'm surprised we would want to do it multiple times on purpose.

**Reply by Kixunil** [reply]
Hmm, true, I guess we open files more often than we add submodules.

**Reply by dpc** [reply]
True. Though personally I [`z`](https://github.com/ajeetdsouza/zoxide) between directories, pick files with [fzf plugin for my shell](https://github.com/PatrickF1/fzf.fish) and generally have a fuzzy search  integrated in every tool I use, so I never notice anymore. And I highly recommend it just because it's a time saver.

**Reply by Kixunil** [reply]
My workflow is quite different but thanks for potential inspiration. :)

**Reply by apoelstra** [reply]
Ok, I will start using `z`, and for now I rescind my complaint :)

**Reply by Kixunil** [reply]
@dpc are you aware of something like `z` but for **files** in neovim? Basically `:e file_name_without_path<tab>` -> `:e some/path/to/file_name_without_path`

**Reply by dpc** [reply]
@Kixunil Sorry, but I've switched from Vim to Kakoune ~6 years ago, and recently became Helix maximalist, so I've lost track of mainstream text editors. :rofl:

**Reply by Kixunil** [reply]
Helix looks interesting, thanks for the tip!

**Reply by Kixunil** [reply]
Speaking of *new* stuff, have you heard about [nushell](https://www.nushell.sh)? :)

**Reply by dpc** [reply]
It's not so nu anymore. I'm planning to switch to it after it starts to feel like Fish Shell has became too mainstream. Ehhh... maintaining Unix Hipster status is a lot of work.

**Reply by tcharding** [reply]
Good god, this is hot AF
```bash
~/b/g/tcharding ❯❯❯ z rust-bitcoin/bitcoin
~/b/g/t/r/bitcoin ❯❯❯ cd
~ ❯❯❯ z bitcoin
~/b/g/t/rust-bitcoin ❯❯❯
```

**Reply by dpc** [reply]
You can do multiple words too! The thing to remember is that decision what you mean depends on the directories you actually `cd`ed in the past. So you still need to `cd` sometimes into places.
